pkg/egressgateway/policy.go

```diff
@@ -177,33 +179,114 @@ func (gwc *gatewayConfig) deriveFromPolicyGatewayConfig(gc *policyGatewayConfig)
 	return nil
 }
 
+// destinationMinusExcludedCIDRs will return, for a given policy, a list of all
+// destination CIDRs to which the excluded CIDRs have been subtracted.
+func (config *PolicyConfig) destinationMinusExcludedCIDRs() []*net.IPNet {
+	if len(config.excludedCIDRs) == 0 {
+		return config.dstCIDRs
+	}
+
+	cidrs := []*net.IPNet{}
+
+	for _, dstCIDR := range config.dstCIDRs {
+		dstCIDRMinusExcludedCIDRs := []*net.IPNet{dstCIDR}
+		for _, excludedCIDR := range config.excludedCIDRs {
+			newDstCIDRMinuxExcludedCIDRs := []*net.IPNet{}
+			for _, cidr := range dstCIDRMinusExcludedCIDRs {
+				r, _, l := ip.PartitionCIDR(*cidr, *excludedCIDR)
+				newDstCIDRMinuxExcludedCIDRs = append(newDstCIDRMinuxExcludedCIDRs, append(r, l...)...)
+			}
+
+			dstCIDRMinusExcludedCIDRs = newDstCIDRMinuxExcludedCIDRs
+		}
+
+		cidrs = append(cidrs, dstCIDRMinusExcludedCIDRs...)
+	}
+
+	return cidrs
+}
+
 func (config *PolicyConfig) forEachEndpointAndDestination(epDataStore map[endpointID]*endpointMetadata,
-	f func(net.IP, *net.IPNet, *gatewayConfig)) {
+	f func(net.IP, *net.IPNet, bool, *gatewayConfig)) {
 
 	for _, endpoint := range epDataStore {
 		if !config.selectsEndpoint(endpoint) {
 			continue
 		}
 
 		for _, endpointIP := range endpoint.ips {
+			isExcludedCIDR := false
 			for _, dstCIDR := range config.dstCIDRs {
-				f(endpointIP, dstCIDR, &config.gatewayConfig)
+				f(endpointIP, dstCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+
+			isExcludedCIDR = true
+			for _, excludedCIDR := range config.excludedCIDRs {
+				f(endpointIP, excludedCIDR, isExcludedCIDR, &config.gatewayConfig)
+			}
+		}
+	}
+}
+
+func (config *PolicyConfig) forEachEndpointAndDestinationMinusExcludedCIDRs(epDataStore map[endpointID]*endpointMetadata,
```

Comment:
What about `forEachEndpointAndDestination` becomes `forEachEndpointAndCIDR` while the new function becomes `forEachEndpointAndComputedDestination` (or just `forEachEndpointAndDestination` if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

In any case, you need comments on all of those helper functions to explain when it uses the input CIDR lists as is vs. the diff of the input CIDRs.

Reply:
> What about forEachEndpointAndDestination becomes forEachEndpointAndCIDR while the new function becomes forEachEndpointAndComputedDestination (or just forEachEndpointAndDestination if that's too long). The rationale being that the first doesn't really process "destinations" anymore but two CIDR lists; the second only processes destination because we computed the diff.

The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L60
https://github.com/cilium/cilium/blob/5510a770cd5c013492df582fcd4caef0eae42e79/pkg/egressgateway/policy.go#L61

(so to me it reads like "iterate over all combinations of EP and destination CIDRs").

Same idea for the other `forEachEndpointAndDestinationMinusExcludedCIDRs` version, which although very mouthful, IMO makes it more clear what we are iterating over: EPs and `dstCIDRs` slice minus `excludedCDIRs` slice.

Perhaps for consistency I can rename the first one to `forEachEndpointAndDestinationCIDRs`

Reply:
> The intent here was to have a self explanatory name, so that I don't have to go and look it up in the comment what it's doing: these are `PolicyConfig`'s methods, so the `Endpoint` and `Destination` part of `forEachEndpointAndDestination` simply refers to the policy's fields:

I think this is too complex to hope that the name can be self-explanatory. I don't think you can avoid adding a comment for those helper functions.

One issue with `forEachEndpointAndDestinationMinusExcludedCIDRs` is that it's unclear, without reading the function itself, if it iterates over all endpoints and destinations, but doesn't consider excluded CIDRs ("minus/except excluded CIDRs") or if it iterates over all endpoints and diffed destinations ("minus/diffed by excluded CIDRs").

Reply:
Fair enough, I'll add some comments then :+1:

Reply:
Scratch that, after taking another fresh look I'm also convinced that `forEachEndpointAndCIDR` and `forEachEndpointAndDestination` are better names, so I went for them :+1:

Reply:
Not to be a pain, but I think a quick comment would still help :slightly_smiling_face:

Reply:
Yes, I added them already in the latest push